File: backend/app/schemas/wallet.py
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, Field

class WalletResponse(BaseModel):
    id: int
    user_id: int
    balance: Decimal
    currency: str
    model_config = {"from_attributes": True}

class WalletTopUpRequest(BaseModel):
    amount: Decimal = Field(gt=0, max_digits=14, decimal_places=2)
    description: str | None = Field(default="Wallet top-up", max_length=255)

class WalletTransactionResponse(BaseModel):
    id: int
    transaction_reference: str
    transaction_type: str
    amount: Decimal
    balance_before: Decimal
    balance_after: Decimal
    status: str
    description: str | None
    booking_id: int | None
    created_at: datetime
    model_config = {"from_attributes": True}

File: frontend/services/wallet.ts
import { api } from "@/lib/api";

export interface Wallet {
  id: number;
  user_id: number;
  balance: number | string;
  currency: string;
}

export interface WalletTransaction {
  id: number;
  transaction_reference: string;
  transaction_type: string;
  amount: number | string;
  balance_before: number | string;
  balance_after: number | string;
  status: string;
  description?: string | null;
  booking_id?: number | null;
  created_at: string;
}

export async function getWallet() {
  const { data } = await api.get<Wallet>("/wallet");
  return data;
}

export async function getWalletTransactions() {
  const { data } = await api.get<WalletTransaction[]>("/wallet/transactions");
  return data;
}

export async function topUpWallet(amount: number, description = "Wallet top-up") {
  const { data } = await api.post<WalletTransaction>("/wallet/top-up", {
    amount,
    description,
  });
  return data;
}
